test/test_ops.py

```diff
@@ -1070,17 +1074,31 @@ def set_requires_grad(x):
                 x.requires_grad_(True)
             return x
 
-        out = pytree.tree_map_(set_requires_grad, expect)
+        pytree.tree_map_(set_requires_grad, out)
         if not any_requires_grad:
             # Skip ops without any floating point outputs, e.g. isnan
             return
 
-        msg = (
-            "functions with out=... arguments don't support automatic "
-            "differentiation, but one of the arguments requires grad."
-        )
-        with self.assertRaises(RuntimeError, msg=msg):
-            op(sample.input, *sample.args, **sample.kwargs, out=out)
+        if not op.supports_autograd_on_out:
+            msg = (
+                "functions with out=... arguments don't support automatic "
+                "differentiation, but one of the arguments requires grad."
+            )
+            with self.assertRaises(RuntimeError, msg=msg):
+                op(sample.input, *sample.args, **sample.kwargs, out=out)
+
+        else:
+            result = op(sample.input, *sample.args, **sample.kwargs, out=out)
+            self.assertEqual(result, None)
+            self.assertIsInstance(out, tuple)
+            self.assertNotEqual(out, ())
+
+            [o.backward() for o in out]
+            self.assertTrue(all(o.grad is not None for o in out))
+
+            if op.name == "unbind_copy":  # Only such op right now
+                self.assertEqual({o.shape for o in out}, {out[0].shape})
+                self.assertTrue(all(torch.all(o.grad == 1.0) for o in out))
```

Comment:
This does not cover many interesting cases for derivative support, so code testing this situation would be best added to `test_ops_gradients.py`. Doing so should allow you to also check the complex dtypes as inputs for backward will be generated correctly. Here you are relying on the coincidental fact that the first sample for `unbind_copy` produces scalar outputs. 

However, getting the out overload to work with gradcheck will be difficult, if not impossible, that kind of reinforces my suspicions that this should not support gradients at all.

Reply:
All right, I went back to the previous state without autogenerating the code, and made sure by hand that we're emitting the right message.